larpandoracontent/LArControlFlow/TrackDirectionTool.cc

```diff
@@ -0,0 +1,1154 @@
+      LArPfoHelper::GetSlidingFitTrajectory(pPfo, pVertex, m_slidingFitWindow, slidingFitPitch, trackStateVector);
+
+     
+
+      const Cluster *const pClusterW = GetTargetClusterFromPFO(pPfo, trackStateVector);
+
+      CaloHitList totalcalohits;
+      LArPfoHelper::GetCaloHits(pPfo, TPC_VIEW_U, totalcalohits);
+      LArPfoHelper::GetCaloHits(pPfo, TPC_VIEW_V, totalcalohits);
+      LArPfoHelper::GetCaloHits(pPfo, TPC_VIEW_W, totalcalohits);
+
+
+        if (pClusterW->GetNCaloHits() <= m_minClusterCaloHits)
+        {
+            throw StatusCodeException(STATUS_CODE_NOT_FOUND);
+        }
+
+        DirectionFitObject finalDirectionFitObject = GetClusterDirection(pClusterW);
+        this->ComputeProbability(finalDirectionFitObject);
+
+        //If the PFO is 3D, then 3D endpoints should be set 
+        if (LArPfoHelper::IsThreeD(pPfo))
+            SetEndpoints(finalDirectionFitObject, trackStateVector);
+
+ 
+        return finalDirectionFitObject;
+    }
+
+    catch (StatusCodeException &statusCodeException)
+    {
+        throw statusCodeException;
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TrackDirectionTool::WriteLookupTableToTree(LookupTable &lookupTable)
+{
+    std::vector<int> mapVector1, reverseMapVector2;
+    std::vector<double> mapVector2, reverseMapVector1;
+
+    for (auto &element : lookupTable.GetMap())
+    {
+        mapVector1.push_back(element.first);
+        mapVector2.push_back(element.second);
+    }
+
+    for (auto &element : lookupTable.GetReverseMap())
+    {
+        reverseMapVector1.push_back(element.first);
+        reverseMapVector2.push_back(element.second);
+    }
+
+    PANDORA_MONITORING_API(SetTreeVariable(this->GetPandora(), m_treeName.c_str(), "mapVector1", &mapVector1));
+    PANDORA_MONITORING_API(SetTreeVariable(this->GetPandora(), m_treeName.c_str(), "mapVector2", &mapVector2));
+    PANDORA_MONITORING_API(SetTreeVariable(this->GetPandora(), m_treeName.c_str(), "reverseMapVector1", &reverseMapVector1));
+    PANDORA_MONITORING_API(SetTreeVariable(this->GetPandora(), m_treeName.c_str(), "reverseMapVector2", &reverseMapVector2));
+    PANDORA_MONITORING_API(SetTreeVariable(this->GetPandora(), m_treeName.c_str(), "maxRange", lookupTable.GetMaxRange()));
+    PANDORA_MONITORING_API(FillTree(this->GetPandora(), m_treeName.c_str()));
+}
+
+  //-----------------------------------------------------------------------------------------------------------------------
+
+const Cluster* TrackDirectionTool::GetTargetClusterFromPFO(const ParticleFlowObject* pPfo, const LArTrackStateVector &trackStateVector)
+{
+    ClusterList clusterListW;
+    LArPfoHelper::GetTwoDClusterList(pPfo, clusterListW);
+
+    if (trackStateVector.size() != 0) {
+      TrackState firstTrackState(*(trackStateVector.begin())), lastTrackState(trackStateVector.back());
```

Comment:
It does not like this: 
![Screenshot from 2021-04-13 11-53-41](https://user-images.githubusercontent.com/50299856/114541672-ef3bfc00-9c4e-11eb-9664-623076562472.png)

Reply:
You don't want to dereference (*) the result of the call to front, as this already gives you back the object, rather than an iterator, so `TrackState firstTrackState(trackStateVector.front())` should resolve this.

Reply:
Ah yes, thanks